src/common/libflux/message.c

```diff
@@ -1247,26 +1247,39 @@ static void userid2str (uint32_t userid, char *buf, int buflen)
     assert (n < buflen);
 }
 
-static void rolemask2str (uint32_t rolemask, char *buf, int buflen)
+static int roletostr (uint32_t role, const char *sep, char *buf, int buflen)
 {
     int n;
-    switch (rolemask) {
-        case FLUX_ROLE_NONE:
-            n = snprintf (buf, buflen, "none");
-            break;
-        case FLUX_ROLE_OWNER:
-            n = snprintf (buf, buflen, "owner");
-            break;
-        case FLUX_ROLE_USER:
-            n = snprintf (buf, buflen, "user");
-            break;
-        case FLUX_ROLE_ALL:
-            n = snprintf (buf, buflen, "all");
-            break;
-        default:
-            n = snprintf (buf, buflen, "unknown");
+    if (role == FLUX_ROLE_OWNER)
+        n = snprintf (buf, buflen, "%sowner", sep);
+    else if (role == FLUX_ROLE_USER)
+        n = snprintf (buf, buflen, "%suser", sep);
+    else if (role == FLUX_ROLE_LOCAL)
+        n = snprintf (buf, buflen, "%slocal", sep);
+    else
+        n = snprintf (buf, buflen, "%s0x%x", sep, role);
+    if (n >= buflen)
+        n = buflen;
+    return n;
+}
+
+static void rolemask2str (uint32_t rolemask, char *buf, int buflen)
+{
+    if (rolemask == FLUX_ROLE_NONE)
+        snprintf (buf, buflen, "none");
+    else if (rolemask == FLUX_ROLE_ALL)
+        snprintf (buf, buflen, "all");
+    else {
+        int offset = 0;
+        for (int i = 0; i < sizeof (rolemask)*8; i++) {
+            if (rolemask & 1<<i) {
+                offset += roletostr (rolemask & 1<<i,
+                                     i > 0 ? "," : "",
```

Comment:
should this be `offset > 0`, only output comma if something was output previously?  Dunno if unit tests possible.

Reply:
Oh you're right, thanks!

Reply:
The magic of [open_memstream(3)](https://man7.org/linux/man-pages/man3/open_memstream.3.html) should make a test possible!